compiler/optimizations/inferConstRefs.cpp

```diff
@@ -366,43 +316,263 @@ static bool inferConstRef(Symbol* sym) {
+  // update any properties of this symbol, and should instead let the first
+  // call on this symbol update such properties.
+  bool isFirstCall = false;
+  if (info && info->alreadyCalled == false) {
+    isFirstCall = true;
+    info->alreadyCalled = true;
+  }
+
+  bool retval = true;
+  int numDefs = 0;
+
+  while (info && info->hasMore() && retval) {
+    SymExpr* use = info->next();
+
+    CallExpr* call = toCallExpr(use->parentExpr);
+    if (call == NULL) {
+      // Could be a DefExpr, or the condition for a while loop.
+      // BHARSH: I'm not sure of all the possibilities
+      continue;
+    }
+
+    CallExpr* parent = toCallExpr(call->parentExpr);
+
+    if (call->isResolved()) {
+      ArgSymbol* form = actual_to_formal(use);
+      if (form->isRef()) {
+        if (!inferConstRef(form)) {
+          retval = false;
+        }
+      }
+    }
+    else if (parent && isMoveOrAssign(parent)) {
+      if (call->isPrimitive(PRIM_ADDR_OF) ||
+          call->isPrimitive(PRIM_SET_REFERENCE)) {
+        SymExpr* LHS = toSymExpr(parent->get(1));
+        if (LHS->isRef() && !inferConstRef(LHS->symbol())) {
+          retval = false;
+        }
+      }
+    }
+    else if (isMoveOrAssign(call)) {
+      if (use == call->get(1)) {
+        numDefs += 1;
+        if (numDefs > 1) {
+          retval = false;
+        }
+      }
     } else {
-      INT_ASSERT(isVarSymbol(sym));
-      sym->qual = QUAL_CONST_REF;
+      // To be safe, exit the loop with 'false' if we're unsure of how to
+      // handle a primitive.
+      retval = false;
     }
   }
 
-  // TODO: assert that it's not already finalized?
-  if (info) {
-    info->finalized = true;
+  if (isFirstCall) {
+    if (retval && !info->finalizedConstness) {
+      if (ArgSymbol* arg = toArgSymbol(sym)) {
+        INT_ASSERT(arg->intent & INTENT_FLAG_IN);
+        arg->intent = INTENT_CONST_IN;
+      } else {
+        INT_ASSERT(isVarSymbol(sym));
+        sym->qual = QUAL_CONST_VAL;
+      }
+    }
+
+    if (info) {
+      info->reset();
+      info->finalizedConstness = true;
+    }
+  } else if (!retval && info) {
+    // Not the first call, but something happened that prevents us from marking
+    // this symbol as const. Now that we know, indicate that this is finalized.
+    info->finalizedConstness = true;
+  }
+
+  return retval;
+}
+
+static bool inferRefToConst(Symbol* sym) {
+  if (!sym->isRef()) return false;
+
+  bool isConstRef = sym->qualType().getQual() == QUAL_CONST_REF;
+  const bool wasRefToConst = sym->hasFlag(FLAG_REF_TO_CONST);
+
+  ConstInfo* info = infoMap[sym];
+
+  // If this ref isn't const, then it can't point to a const thing
+  if ((info && info->finalizedRefToConst) || wasRefToConst || !isConstRef) {
+    return wasRefToConst;
+  }
+
+  bool isFirstCall = false;
+  if (info && info->alreadyCalled == false) {
+    isFirstCall = true;
+    info->alreadyCalled = true;
+  }
+
+  bool retval = true;
```

Comment:
here, and in other (long) functions, I think it'd be nicer to call it, say 'isRefToConst' than 'retval'.